### PR TITLE
feat(router): add generic merchant static routing config

### PR DESCRIPTION
**Pull request by vspecky**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates

## Description

This PR removes the legacy `CustomRoutingRules` type for describing static routing and makes the static routing configuration in the merchant account generic.
It also adds a simple `Single` static routing algorithm which will allow the merchant to route all payments to a single connector and its implementation in payments core.

### Additional Changes

- [x] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

- [DB up.sql](https://github.com/juspay/hyperswitch/blob/636c758f67dd14ea8735bf05f1635002cb424a36/migrations/2023-01-12-084710_update_merchant_routing_algorithm/up.sql)
- [DB down.sql](https://github.com/juspay/hyperswitch/blob/636c758f67dd14ea8735bf05f1635002cb424a36/migrations/2023-01-12-084710_update_merchant_routing_algorithm/down.sql)

## Motivation and Context

This serves as scaffolding to integrate other forms of static routing into hyperswitch.

## How did you test it?

Compiler-guided, stripe payments.

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Comment by vspecky** [comment]
We need to have a discussion on the implementation for the intersection of the following aspects of payments core and make necessary changes before merging the PR :-
- Static Routing
- Registering tasks in process-tracker
- Automatic retries

**Review by SanchithHegde — COMMENTED** [review]
Other than that, looks good to me!

File: crates/api_models/src/admin.rs [hunk 1]
```diff
@@ -32,12 +32,8 @@ pub struct CreateMerchantAccount {
     /// Webhook related details
     pub webhook_details: Option<WebhookDetails>,
 
-    /// The routing algorithm to be used to process the incoming request from merchant to outgoing payment processor or payment method. The default is 'Custom'
-    #[schema(value_type = Option<RoutingAlgorithm>, max_length = 255, example = "custom")]
-    pub routing_algorithm: Option<api_enums::RoutingAlgorithm>,
-
     /// The custom routing rules to be used for various payment methods and conditions
-    pub custom_routing_rules: Option<Vec<CustomRoutingRules>>,
+    pub routing_algorithm: Option<serde_json::Value>,
```
[the remark this answers is not in the document]
Comment (vspecky) [reply crates/api_models/src/admin.rs:36]: Yes, better if we refer to it as "routing algorithm config" in the doc.

File: crates/api_models/src/enums.rs [hunk 2]
```diff
@@ -513,6 +513,34 @@ pub enum Connector {
...
+    Authorizedotnet,
+    Braintree,
+    Checkout,
+    Cybersource,
+    Globalpay,
+    Klarna,
+    Shift4,
+    Stripe,
+    Worldpay,
+}
```
[the remark this answers is not in the document]
Comment (vspecky) [reply crates/api_models/src/enums.rs:542]: Connectors like `Applepay` are not connectors in the sense that merchants can call them to make payments. These are used internally. That's the purpose of this enum, it only enumerates merchant-callable connectors.

File: crates/router/src/core/payments/helpers.rs [hunk 3]
```diff
@@ -594,7 +593,7 @@ pub async fn get_customer_from_details(
 }
 
 pub async fn get_connector_default(
-    merchant_account: &storage::MerchantAccount,
+    _merchant_account: &storage::MerchantAccount,
```
[the remark this answers is not in the document]
Comment (vspecky) [reply crates/router/src/core/payments/helpers.rs:596]: Might need in the future but no need to keep it now. Will remove.

File: crates/api_models/src/enums.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (vspecky) [reply crates/api_models/src/enums.rs:542]: I believe those are merchant-callable. Will update.

[the remark this answers is not in the document]
Comment (vspecky) [reply crates/api_models/src/enums.rs:542]: I guess the name could also be better. `SupportedConnectors` doesn't really drive home the purpose. Maybe something like `MerchantCallableConnectors`. Too long tho, I guess `RoutableConnectors` fits.

File: crates/api_models/src/admin.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (SanchithHegde) [reply crates/api_models/src/admin.rs:36]: > ```rust
> /// The custom routing rules to be used for various payment methods and conditions
> pub routing_algorithm: Option<serde_json::Value>,
> ```

Should the doc comment say "custom routing rules", or would you prefer to update it?

**Review by SanchithHegde — DISMISSED** [review]
Looks good to me!

File: crates/router/src/core/payments.rs [hunk 4]
```diff
@@ -105,6 +105,16 @@ where
     if let api::ConnectorCallType::Single(ref connector) = connector_details {
```
Comment (jarnura) [inline crates/router/src/core/payments.rs:105]: move this logic to a separate function.

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payments.rs (hunk 4 above)
Comment (vspecky) [reply crates/router/src/core/payments.rs]: Done 👍

**Review by SanchithHegde — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**